---
title: ç»„ä»¶å’Œ Hook å¿…é¡»æ˜¯çº¯ç²¹çš„
---

<Intro>
Pure functions only perform a calculation and nothing more. It makes your code easier to understand, debug, and allows React to automatically optimize your components and Hooks correctly.
</Intro>

<Note>
æœ¬å‚è€ƒé¡µé¢æ¶µç›–é«˜çº§è¯é¢˜ï¼Œéœ€è¦ç†Ÿæ‚‰ [ä¿æŒç»„ä»¶çº¯ç²¹](/learn/keeping-components-pure) é¡µé¢ä¸­æ¶‰åŠçš„æ¦‚å¿µã€‚
</Note>

<InlineToc />

### ä¸ºä»€ä¹ˆä¿æŒçº¯ç²¹å¾ˆé‡è¦ï¼Ÿ {/*why-does-purity-matter*/}

React ä¸­çš„ä¸€ä¸ªæ ¸å¿ƒæ¦‚å¿µæ˜¯ä¿æŒçº¯ç²¹ã€‚ä¸€ä¸ªçº¯ç»„ä»¶æˆ– Hook æ˜¯æŒ‡ï¼š

* **Idempotent** â€“ You [always get the same result every time](/learn/keeping-components-pure#purity-components-as-formulas) you run it with the same inputs â€“ props, state, context for component inputs; and arguments for hook inputs.
* **Has no side effects in render** â€“ Code with side effects should run [**separately from rendering**](#how-does-react-run-your-code). For example as an [event handler](/learn/responding-to-events) â€“ where the user interacts with the UI and causes it to update; or as an [Effect](/reference/react/useEffect) â€“ which runs after render.
* **Does not mutate non-local values**: Components and Hooks should [never modify values that aren't created locally](#mutation) in render.

å½“æ¸²æŸ“ä¿æŒçº¯å‡€æ—¶ï¼ŒReact èƒ½å¤Ÿç†è§£å“ªäº›æ›´æ–°å¯¹ç”¨æˆ·æ¥è¯´æœ€é‡è¦ï¼Œåº”è¯¥ä¼˜å…ˆæ˜¾ç¤ºã€‚è¿™æ˜¯å› ä¸ºæ¸²æŸ“çš„çº¯ç²¹ï¼Œå³ç”±äºç»„ä»¶ [åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­](#how-does-react-run-your-code) ä¸ä¼šäº§ç”Ÿå‰¯ä½œç”¨ï¼ŒReact å¯ä»¥æš‚åœæ¸²æŸ“é‚£äº›ä¸æ˜¯é‚£ä¹ˆé‡è¦çš„ç»„ä»¶ï¼Œç­‰åˆ°çœŸæ­£éœ€è¦æ—¶å†ç»§ç»­æ¸²æŸ“å®ƒä»¬ã€‚

å…·ä½“æ¥è¯´ï¼Œè¿™æ„å‘³ç€æ¸²æŸ“é€»è¾‘å¯ä»¥å¤šæ¬¡è¿è¡Œï¼Œè¿™æ · React å°±èƒ½å¤Ÿä¸ºä½ çš„ç”¨æˆ·æä¾›æ„‰å¿«çš„ä½“éªŒã€‚ç„¶è€Œï¼Œå¦‚æœä½ çš„ç»„ä»¶ [åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­](#how-does-react-run-your-code) æœ‰æ— è¿½è¸ªçš„å‰¯ä½œç”¨ï¼Œæ¯”å¦‚ä¿®æ”¹å…¨å±€å˜é‡çš„å€¼ï¼Œé‚£ä¹ˆå½“ React å†æ¬¡è¿è¡Œä½ çš„æ¸²æŸ“ä»£ç æ—¶ï¼Œè¿™äº›å‰¯ä½œç”¨ä¼šä»¥ä½ ä¸å¸Œæœ›çš„æ–¹å¼è¢«è§¦å‘ã€‚è¿™é€šå¸¸ä¼šå¯¼è‡´æ„å¤–çš„ bugï¼Œä»è€Œé™ä½ç”¨æˆ·å¯¹ä½ åº”ç”¨çš„ä½“éªŒæ„Ÿã€‚ä½ å¯ä»¥çœ‹åˆ°è¿™æ ·ä¸€ä¸ª [ä¾‹å­åœ¨ä¿æŒç»„ä»¶çº¯ç²¹é¡µé¢ä¸­](/learn/keeping-components-pure#side-effects-unintended-consequences)ã€‚

#### React æ˜¯å¦‚ä½•è¿è¡Œä½ çš„ä»£ç çš„ï¼Ÿ {/*how-does-react-run-your-code*/}

React æ˜¯å£°æ˜å¼çš„ï¼Œå³ä½ å‘Šè¯‰ React ä½ æƒ³è¦æ¸²æŸ“çš„å†…å®¹ï¼ŒReact ä¼šè‡ªå·±é€‰æ‹©æœ€ä½³çš„æ–¹å¼å‘ç”¨æˆ·å±•ç¤ºå®ƒã€‚ä¸ºäº†åšåˆ°è¿™ä¸€ç‚¹ï¼ŒReact åœ¨æ‰§è¡Œä½ çš„ä»£ç æ—¶åˆ†ä¸ºå‡ ä¸ªé˜¶æ®µã€‚è™½ç„¶ä½ ä¸å¿…äº†è§£æ‰€æœ‰è¿™äº›é˜¶æ®µå°±èƒ½å¾ˆå¥½åœ°ä½¿ç”¨ Reactã€‚ä½†æ˜¯ï¼Œä»é«˜å±‚æ¬¡æ¥çœ‹ï¼Œä½ åº”è¯¥äº†è§£å“ªäº›ä»£ç åœ¨æ¸²æŸ“é˜¶æ®µè¿è¡Œï¼Œå“ªäº›ä»£ç åœ¨æ¸²æŸ“é˜¶æ®µä¹‹å¤–è¿è¡Œã€‚

_Rendering_ refers to calculating what the next version of your UI should look like. After rendering, [Effects](/reference/react/useEffect) are _flushed_ (meaning they are run until there are no more left) and may update the calculation if the Effects have impacts on layout. React takes this new calculation and compares it to the calculation used to create the previous version of your UI, then _commits_ just the minimum changes needed to the [DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) (what your user actually sees) to catch it up to the latest version.

<DeepDive>

#### å¦‚ä½•åˆ¤æ–­ä»£ç æ˜¯å¦åœ¨æ¸²æŸ“ä¸­è¿è¡Œ {/*how-to-tell-if-code-runs-in-render*/}

ä¸€ä¸ªå¿«é€Ÿåˆ¤æ–­ä»£ç æ˜¯å¦åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­è¿è¡Œçš„æ–¹æ³•æ˜¯æ£€æŸ¥ä»£ç çš„ä½ç½®ï¼šå¦‚æœå®ƒåƒä¸‹é¢çš„ä¾‹å­é‚£æ ·å†™åœ¨é¡¶å±‚ï¼Œé‚£ä¹ˆå®ƒå¾ˆå¯èƒ½ä¼šåœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­è¿è¡Œã€‚

```js {2}
function Dropdown() {
  const selectedItems = new Set(); // åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­åˆ›å»º
  // ...
}
```

äº‹ä»¶å¤„ç†å‡½æ•°å’Œ Effect åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­ä¸ä¼šè¿è¡Œï¼š

```js {4}
function Dropdown() {
  const selectedItems = new Set();
  const onSelect = (item) => {
    // è¿™æ®µä»£ç ä½äºäº‹ä»¶å¤„ç†å‡½æ•°ä¸­ï¼Œå› æ­¤åªæœ‰åœ¨ç”¨æˆ·è§¦å‘è¿™ä¸ªäº‹ä»¶æ—¶æ‰ä¼šè¿è¡Œã€‚
    selectedItems.add(item);
  }
}
```

```js {4}
function Dropdown() {
  const selectedItems = new Set();
  useEffect(() => {
    // è¿™æ®µä»£ç ä½äº Effect å†…éƒ¨ï¼Œå› æ­¤å®ƒåªåœ¨æ¸²æŸ“å®Œæˆåè¿è¡Œã€‚
    logForAnalytics(selectedItems);
  }, [selectedItems]);
}
```
</DeepDive>

---

## Components and Hooks must be idempotent {/*components-and-hooks-must-be-idempotent*/}

Components must always return the same output with respect to their inputs â€“ props, state, and context. This is known as _idempotency_. [Idempotency](https://en.wikipedia.org/wiki/Idempotence) is a term popularized in functional programming. It refers to the idea that you [always get the same result every time](learn/keeping-components-pure) you run that piece of code with the same inputs.

è¿™æ„å‘³ç€ï¼Œä¸ºäº†éµå¾ªè¿™ä¸€è§„åˆ™ï¼Œæ‰€æœ‰ [åœ¨æ¸²æŸ“æœŸé—´](#how-does-react-run-your-code)  æ‰§è¡Œçš„ä»£ç ä¹Ÿå¿…é¡»æ˜¯å¹‚ç­‰çš„ã€‚ä¾‹å¦‚ï¼Œä»¥ä¸‹è¿™è¡Œä»£ç å°±ä¸æ˜¯å¹‚ç­‰çš„ï¼ˆå› æ­¤ï¼ŒåŒ…å«è¿™è¡Œä»£ç çš„ç»„ä»¶ä¹Ÿä¸æ˜¯å¹‚ç­‰çš„ï¼‰ï¼š

```js {2}
function Clock() {
  const time = new Date(); // ğŸ”´ é”™è¯¯çš„ï¼šæ€»æ˜¯è¿”å›ä¸åŒçš„ç»“æœï¼
  return <span>{time.toLocaleString()}</span>
}
```

`new Date()` å‡½æ•°ä¸æ˜¯å¹‚ç­‰çš„ï¼Œå› ä¸ºå®ƒæ€»æ˜¯è¿”å›å½“å‰çš„æ—¥æœŸå’Œæ—¶é—´ï¼Œå¹¶ä¸”æ¯æ¬¡è°ƒç”¨æ—¶è¿”å›çš„ç»“æœéƒ½ä¸åŒã€‚å½“ä½ æ¸²æŸ“ä¸Šé¢çš„ç»„ä»¶æ—¶ï¼Œå±å¹•ä¸Šæ˜¾ç¤ºçš„æ—¶é—´å°†ä¼šåœç•™åœ¨ç»„ä»¶è¢«æ¸²æŸ“çš„é‚£ä¸€åˆ»çš„æ—¶é—´ã€‚ç±»ä¼¼åœ°ï¼Œåƒ `Math.random()` è¿™æ ·çš„å‡½æ•°ä¹Ÿä¸æ˜¯å¹‚ç­‰çš„ï¼Œå› ä¸ºå³ä½¿è¾“å…¥ç›¸åŒï¼Œå®ƒä»¬æ¯æ¬¡è°ƒç”¨ä¹Ÿéƒ½ä¼šè¿”å›ä¸åŒçš„ç»“æœã€‚

è¿™å¹¶ä¸æ„å‘³ç€ä½ å®Œå…¨ä¸èƒ½ä½¿ç”¨åƒ `new Date()` è¿™æ ·éå¹‚ç­‰çš„å‡½æ•°â€”â€”ä½ åªéœ€è¦é¿å… [åœ¨æ¸²æŸ“è¿‡ç¨‹](#how-does-react-run-your-code) ä¸­ä½¿ç”¨å®ƒä»¬å³å¯ã€‚åœ¨è¿™ç§æƒ…å†µä¸‹ï¼Œæˆ‘ä»¬å¯ä»¥ä½¿ç”¨ä¸€ä¸ª [Effect](/reference/react/useEffect) æ¥å°†æœ€æ–°çš„æ—¥æœŸä¸è¿™ä¸ªç»„ä»¶è¿›è¡Œâ€œåŒæ­¥â€ï¼š

<Sandpack>

```js
import { useState, useEffect } from 'react';

function useTime() {
  // 1. è·Ÿè¸ªå½“å‰æ—¥æœŸçš„çŠ¶æ€ã€‚`useState` æ¥å—ä¸€ä¸ªåˆå§‹åŒ–å‡½æ•°ä½œä¸ºå…¶
  //    åˆå§‹çŠ¶æ€ã€‚å®ƒåªåœ¨è°ƒç”¨ Hook æ—¶è¿è¡Œä¸€æ¬¡ï¼Œå› æ­¤åªæœ‰è°ƒç”¨ Hook æ—¶çš„
  //    å½“å‰æ—¥æœŸæ‰è¢«é¦–å…ˆè®¾ç½®ã€‚
  const [time, setTime] = useState(() => new Date());

  useEffect(() => {
    // 2. ä½¿ç”¨ `setInterval` æ¯ç§’æ›´æ–°å½“å‰æ—¥æœŸã€‚
    const id = setInterval(() => {
      setTime(new Date()); // âœ… æ­£ç¡®çš„ï¼šéå¹‚ç­‰ä»£ç ä¸å†åœ¨æ¸²æŸ“ä¸­è¿è¡Œã€‚
    }, 1000);
    // 3. è¿”å›ä¸€ä¸ªæ¸…ç†å‡½æ•°ï¼Œè¿™æ ·æˆ‘ä»¬å°±ä¸ä¼šå¿˜è®°æ¸…ç† `setInterval` å®šæ—¶å™¨ï¼Œå¯¼è‡´å†…å­˜æ³„æ¼ã€‚
    return () => clearInterval(id);
  }, []);

  return time;
}

export default function Clock() {
  const time = useTime();
  return <span>{time.toLocaleString()}</span>;
}
```

</Sandpack>

é€šè¿‡å°†éå¹‚ç­‰çš„ `new Date()` è°ƒç”¨åŒ…è£…åœ¨ä¸€ä¸ª Effect ä¸­ï¼Œå°±å¯ä»¥å°†è¿™ä¸ªè®¡ç®—ç§»åŠ¨åˆ° [æ¸²æŸ“ä¹‹å¤–](#how-does-react-run-your-code)ã€‚

å¦‚æœä½ ä¸éœ€è¦å°†æŸäº›å¤–éƒ¨çŠ¶æ€ä¸ React åŒæ­¥ï¼Œåªéœ€è¦åœ¨å“åº”ç”¨æˆ·äº¤äº’æ—¶æ›´æ–°ï¼Œä½ å¯ä»¥è€ƒè™‘ä½¿ç”¨ä¸€ä¸ª [äº‹ä»¶å¤„ç†å‡½æ•°](/learn/responding-to-events)ã€‚

---

## å‰¯ä½œç”¨å¿…é¡»åœ¨æ¸²æŸ“ä¹‹å¤–æ‰§è¡Œ {/*side-effects-must-run-outside-of-render*/}

[å‰¯ä½œç”¨](/learn/keeping-components-pure#side-effects-unintended-consequences) ä¸åº”è¯¥ [åœ¨æ¸²æŸ“ä¸­](#how-does-react-run-your-code) æ‰§è¡Œï¼Œå› ä¸º React å¯èƒ½ä¼šå¤šæ¬¡æ¸²æŸ“ç»„ä»¶ä»¥æä¾›æœ€ä½³çš„ç”¨æˆ·ä½“éªŒã€‚

<Note>
å‰¯ä½œç”¨æ˜¯ä¸€ä¸ªæ¯” Effect æ›´å¹¿æ³›çš„æ¦‚å¿µã€‚Effect ç‰¹æŒ‡è¢«åŒ…è£¹åœ¨ `useEffect` ä¸­çš„ä»£ç ï¼Œè€Œâ€œå‰¯ä½œç”¨â€æ˜¯ä¸€èˆ¬æœ¯è¯­ï¼ŒæŒ‡é™¤äº†å°†å…¶ä¸»è¦ç»“æœï¼ˆè¿”å›å€¼ï¼‰ä¼ é€’ç»™è°ƒç”¨è€…ä¹‹å¤–ï¼Œå¯¹å¤–éƒ¨ä¸–ç•Œæœ‰ä»»ä½•å¯è§‚å¯Ÿå½±å“çš„ä»£ç ã€‚

å‰¯ä½œç”¨é€šå¸¸å†™åœ¨ [äº‹ä»¶å¤„ç†å‡½æ•°](/learn/responding-to-events) æˆ– Effect å†…éƒ¨ã€‚ä½†ç»ä¸èƒ½åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­å†™ã€‚
</Note>

å°½ç®¡æ¸²æŸ“å¿…é¡»ä¿æŒçº¯å‡€ï¼Œä½†å‰¯ä½œç”¨å¯¹äºä½ çš„åº”ç”¨æ¥è¯´æ˜¯åº”å½“ä¹Ÿæ˜¯éå¸¸å¿…è¦çš„ï¼Œè¿™æ ·æ‰èƒ½åšä¸€äº›æœ‰è¶£çš„äº‹æƒ…ï¼Œæ¯”å¦‚åœ¨å±å¹•ä¸Šæ˜¾ç¤ºå†…å®¹ï¼è¿™æ¡è§„åˆ™çš„å…³é”®ç‚¹åœ¨äºï¼Œå‰¯ä½œç”¨ä¸åº”è¯¥ [åœ¨æ¸²æŸ“ä¸­](#how-does-react-run-your-code) æ‰§è¡Œï¼Œå› ä¸º React å¯èƒ½ä¼šå¤šæ¬¡æ¸²æŸ“ç»„ä»¶ã€‚åœ¨å¤§å¤šæ•°æƒ…å†µä¸‹ï¼Œä½ ä¼šä½¿ç”¨ [äº‹ä»¶å¤„ç†å‡½æ•°](learn/responding-to-events) æ¥å¤„ç†å‰¯ä½œç”¨ã€‚ä½¿ç”¨äº‹ä»¶å¤„ç†å‡½æ•°æ˜ç¡®åœ°å‘Šè¯‰ React è¿™æ®µä»£ç ä¸éœ€è¦åœ¨æ¸²æŸ“è¿‡ç¨‹ä¸­æ‰§è¡Œï¼Œä»è€Œä¿æŒæ¸²æŸ“çš„çº¯ç²¹ã€‚å¦‚æœä½ å·²ç»å°è¯•äº†æ‰€æœ‰å¯èƒ½çš„æ–¹æ³•â€”â€”å¹¶ä¸”åªæ˜¯ä½œä¸ºæœ€åçš„è§£å†³åŠæ³•â€”â€”ä½ ä¹Ÿå¯ä»¥ä½¿ç”¨ `useEffect` æ¥å¤„ç†å‰¯ä½œç”¨ã€‚

### ä»€ä¹ˆæ—¶å€™å¯ä»¥è¿›è¡Œ mutationï¼Ÿ {/*mutation*/}

#### å±€éƒ¨ mutation {/*local-mutation*/}
ä¸€ä¸ªå¸¸è§çš„å‰¯ä½œç”¨ç¤ºä¾‹æ˜¯çªå˜ï¼ˆmutationï¼‰ï¼Œè¿™åœ¨ JavaScript ä¸­æŒ‡çš„æ˜¯æ”¹å˜ä¸€ä¸ªé [åŸå§‹å€¼](https://developer.mozilla.org/en-US/docs/Glossary/Primitive) çš„å€¼ã€‚é€šå¸¸æ¥è¯´ï¼Œåœ¨ React ä¸­ mutation æ“ä½œå¹¶ä¸ç¬¦åˆæœ€ä½³å®è·µï¼Œä½†æ˜¯è¿›è¡Œå±€éƒ¨ mutation æ˜¯å®Œå…¨å¯ä»¥æ¥å—çš„ï¼š

```js {2,7}
function FriendList({ friends }) {
  const items = []; // âœ… æ­£ç¡®çš„ï¼šåœ¨å±€éƒ¨åˆ›å»º
  for (let i = 0; i < friends.length; i++) {
    const friend = friends[i];
    items.push(
      <Friend key={friend.id} friend={friend} />
    ); // âœ… æ­£ç¡®çš„ï¼šå±€éƒ¨ä¿®æ”¹æ˜¯å¯ä»¥çš„ã€‚
  }
  return <section>{items}</section>;
}
```

ä½ æ²¡æœ‰å¿…è¦ä¸ºäº†å›é¿å±€éƒ¨ mutation è€Œåˆ»æ„ç¼–å†™å¤æ‚çš„ä»£ç ã€‚è™½ç„¶ä¸ºäº†ç®€æ´ï¼Œè¿™é‡Œå¯ä»¥ä½¿ç”¨ [`Array.map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)ï¼Œä½†åˆ›å»ºä¸€ä¸ªå±€éƒ¨æ•°ç»„ï¼Œç„¶å [åœ¨æ¸²æŸ“æ—¶](#how-does-react-run-your-code) å‘å…¶ä¸­æ·»åŠ æ•°ç»„é¡¹ä¹Ÿæ˜¯å®Œå…¨å¯ä»¥çš„ã€‚

å°½ç®¡çœ‹èµ·æ¥æˆ‘ä»¬æ­£åœ¨ä¿®æ”¹ `items`ï¼Œä½†å…³é”®çš„ä¸€ç‚¹æ˜¯è¿™ç§ mutation æ˜¯å±€éƒ¨çš„ï¼Œå½“ç»„ä»¶å†æ¬¡æ¸²æŸ“æ—¶ï¼Œè¿™ç§ mutation ä¸ä¼šè¢«â€œè®°ä½â€ã€‚æ¢å¥è¯è¯´ï¼Œ`items` åªåœ¨ç»„ä»¶å­˜åœ¨æœŸé—´æœ‰æ•ˆã€‚å› ä¸ºæ¯æ¬¡æ¸²æŸ“ `<FriendList />` æ—¶ï¼Œ`items` éƒ½ä¼šè¢«é‡æ–°åˆ›å»ºï¼Œæ‰€ä»¥ç»„ä»¶æ€»èƒ½è¿”å›ç›¸åŒçš„ç»“æœã€‚

å¦ä¸€æ–¹é¢ï¼Œå¦‚æœ `items` æ˜¯åœ¨ç»„ä»¶å¤–éƒ¨åˆ›å»ºçš„ï¼Œé‚£ä¹ˆå®ƒä¼šä¿ç•™å…¶ä¹‹å‰çš„å€¼ï¼Œå¹¶è®°ä½æ‰€åšçš„æ›´æ”¹ï¼š

```js {1,7}
const items = []; // ğŸ”´ é”™è¯¯çš„ï¼šåœ¨ç»„ä»¶å¤–éƒ¨åˆ›å»º
function FriendList({ friends }) {
  for (let i = 0; i < friends.length; i++) {
    const friend = friends[i];
    items.push(
      <Friend key={friend.id} friend={friend} />
    ); // ğŸ”´ é”™è¯¯çš„ï¼šä¿®æ”¹äº†ä¸€ä¸ªåœ¨æ¸²æŸ“ä¹‹å¤–åˆ›å»ºçš„å€¼ã€‚
  }
  return <section>{items}</section>;
}
```

æ¯å½“ `<FriendList />` ç»„ä»¶å†æ¬¡è¿è¡Œæ—¶ï¼Œæˆ‘ä»¬éƒ½ä¼šæŒç»­åœ°å‘ `items` æ•°ç»„è¿½åŠ  `friends`ï¼Œè¿™å°†å¯¼è‡´äº§ç”Ÿå¤šä¸ªé‡å¤çš„ç»“æœã€‚è¿™ä¸ªç‰ˆæœ¬çš„ `<FriendList />` [åœ¨æ¸²æŸ“ä¸­](#how-does-react-run-your-code) å…·æœ‰å¯è§‚å¯Ÿçš„å‰¯ä½œç”¨ï¼Œæ‰€ä»¥è¿åäº†è§„åˆ™ã€‚

#### å»¶è¿Ÿåˆå§‹åŒ– {/*lazy-initialization*/}

å³ä½¿ä¸æ˜¯å®Œå…¨â€œçº¯ç²¹â€çš„ï¼Œå»¶è¿Ÿåˆå§‹åŒ–ä¹Ÿæ˜¯å®Œå…¨å¯ä»¥æ¥å—çš„ï¼š

```js {2}
function ExpenseForm() {
  SuperCalculator.initializeIfNotReady(); // âœ… æ­£ç¡®çš„ï¼šå¦‚æœå®ƒå¯¹å…¶ä»–ç»„ä»¶æ²¡æœ‰å½±å“ã€‚
  // Continue rendering...
}
```

#### æ”¹å˜ DOM {/*changing-the-dom*/}

åœ¨ React ç»„ä»¶çš„æ¸²æŸ“é€»è¾‘ä¸­ä¸å…è®¸æœ‰ç›´æ¥å¯¹ç”¨æˆ·å¯è§çš„å‰¯ä½œç”¨ã€‚æ¢å¥è¯è¯´ï¼Œä»…ä»…è°ƒç”¨ä¸€ä¸ªç»„ä»¶å‡½æ•°æœ¬èº«ä¸åº”å½“åœ¨å±å¹•ä¸Šäº§ç”Ÿå˜åŒ–ã€‚

```js {2}
function ProductDetailPage({ product }) {
  document.window.title = product.title; // ğŸ”´ é”™è¯¯çš„ï¼šæ”¹å˜ DOM
}
```

è¦åœ¨æ¸²æŸ“ä¹‹å¤–æ›´æ–° `window.title` çš„ä¸€ä¸ªæ–¹æ³•æ˜¯ [å°†ç»„ä»¶ä¸ `window` è¿›è¡ŒåŒæ­¥](/learn/synchronizing-with-effects)ã€‚

åªè¦å¤šæ¬¡è°ƒç”¨ç»„ä»¶æ˜¯å®‰å…¨çš„ï¼Œå¹¶ä¸”ä¸ä¼šå½±å“å…¶ä»–ç»„ä»¶çš„æ¸²æŸ“ï¼ŒReact å°±ä¸ä¼šåœ¨æ„ç»„ä»¶æ˜¯å¦åœ¨ä¸¥æ ¼çš„å‡½æ•°å¼ç¼–ç¨‹æ„ä¹‰ä¸Šæ˜¯ç™¾åˆ†ä¹‹ç™¾çº¯ç²¹çš„ã€‚æ›´é‡è¦çš„æ˜¯ï¼Œ[ç»„ä»¶å¿…é¡»æ˜¯å¹‚ç­‰çš„](/reference/rules/components-and-hooks-must-be-pure)ã€‚

---

## props å’Œ state æ˜¯ä¸å¯å˜çš„ {/*props-and-state-are-immutable*/}

ç»„ä»¶çš„ props å’Œ state æ˜¯ä¸å¯å˜çš„ [å¿«ç…§](learn/state-as-a-snapshot)ã€‚æ°¸è¿œä¸è¦ç›´æ¥ä¿®æ”¹å®ƒä»¬ã€‚ç›¸åï¼Œä½ åº”è¯¥å‘ä¸‹ä¼ é€’æ–°çš„å±æ€§ï¼Œä»¥åŠä½¿ç”¨ `useState` æä¾›çš„ setter å‡½æ•°ã€‚

ä½ å¯ä»¥å°† props å’Œ state è§†ä¸ºåœ¨æ¸²æŸ“åæ›´æ–°çš„å¿«ç…§ã€‚å› æ­¤ï¼Œä½ ä¸ä¼šç›´æ¥ä¿®æ”¹ props æˆ– stateï¼Œç›¸åï¼Œä½ ä¼ é€’æ–°çš„ propsï¼Œæˆ–è€…ä½¿ç”¨æä¾›ç»™ä½ çš„ setter å‡½æ•°æ¥å‘Šè¯‰ Reactï¼Œstate éœ€è¦åœ¨ä¸‹ä¸€æ¬¡ç»„ä»¶æ¸²æŸ“æ—¶æ›´æ–°ã€‚

### ä¸è¦ä¿®æ”¹ props {/*props*/}
props æ˜¯ä¸å¯å˜çš„ï¼Œå› ä¸ºå¦‚æœä½ æ”¹å˜äº†å®ƒä»¬ï¼Œåº”ç”¨ç¨‹åºå¯èƒ½ä¼šäº§ç”Ÿä¸ä¸€è‡´çš„ç»“æœï¼Œè¿™ä¼šè®©è°ƒè¯•å˜å¾—å›°éš¾ï¼Œå› ä¸ºç¨‹åºå¯èƒ½ä¼šåœ¨æŸäº›æƒ…å†µä¸‹å·¥ä½œï¼Œè€Œåœ¨å¦ä¸€äº›æƒ…å†µä¸‹ä¸å·¥ä½œã€‚

```js {2}
function Post({ item }) {
  item.url = new Url(item.url, base); // ğŸ”´ é”™è¯¯çš„ï¼šæ°¸è¿œä¸è¦ç›´æ¥ä¿®æ”¹ props
  return <Link url={item.url}>{item.title}</Link>;
}
```

```js {2}
function Post({ item }) {
  const url = new Url(item.url, base); // âœ… æ­£ç¡®çš„ï¼šåˆ›å»ºä¸€ä¸ªæ–°çš„å‰¯æœ¬æ›¿ä»£
  return <Link url={url}>{item.title}</Link>;
}
```

### ä¸è¦ä¿®æ”¹ state {/*state*/}
`useState` è¿”å›ä¸€ä¸ª state å’Œä¸€ä¸ªç”¨äºæ›´æ–°è¯¥çŠ¶æ€çš„ setterã€‚

```js
const [stateVariable, setter] = useState(0);
```

æˆ‘ä»¬ä¸åº”è¯¥ç›´æ¥åœ¨ state å˜é‡ä¸Šè¿›è¡Œæ›´æ–°ï¼Œè€Œåº”è¯¥ä½¿ç”¨ `useState` è¿”å›çš„ setter å‡½æ•°æ¥è¿›è¡Œæ›´æ–°ã€‚å¦‚æœåœ¨ state å˜é‡ä¸Šç›´æ¥ä¿®æ”¹å€¼ï¼Œå¹¶ä¸ä¼šå¯¼è‡´ç»„ä»¶ç•Œé¢æ›´æ–°ï¼Œè¿™æ ·ç”¨æˆ·ç•Œé¢å°±ä¼šæ˜¾ç¤ºè¿‡æ—¶çš„ä¿¡æ¯ã€‚é€šè¿‡ä½¿ç”¨ setter å‡½æ•°ï¼Œæˆ‘ä»¬å‘Šè¯‰ React çŠ¶æ€å·²ç»å‘ç”Ÿäº†å˜åŒ–ï¼Œéœ€è¦è¿›è¡Œé‡æ–°æ¸²æŸ“ï¼Œä»¥ä¾¿æ›´æ–°ç”¨æˆ·ç•Œé¢ã€‚

```js {5}
function Counter() {
  const [count, setCount] = useState(0);

  function handleClick() {
    count = count + 1; // ğŸ”´ é”™è¯¯çš„ï¼šæ°¸è¿œä¸è¦ç›´æ¥ä¿®æ”¹ state
  }

  return (
    <button onClick={handleClick}>
      You pressed me {count} times
    </button>
  );
}
```

```js {5}
function Counter() {
  const [count, setCount] = useState(0);

  function handleClick() {
    setCount(count + 1); // âœ… æ­£ç¡®çš„ï¼šä½¿ç”¨ç”± useState è¿”å›çš„ setter å‡½æ•°æ¥ä¿®æ”¹ stateã€‚
  }

  return (
    <button onClick={handleClick}>
      You pressed me {count} times
    </button>
  );
}
```

---

## Hook çš„è¿”å›å€¼å’Œå‚æ•°æ˜¯ä¸å¯å˜çš„ {/*return-values-and-arguments-to-hooks-are-immutable*/}

ä¸€æ—¦å€¼è¢«ä¼ é€’ç»™ Hookï¼Œå°±ä¸åº”è¯¥å†å¯¹å®ƒä»¬è¿›è¡Œä¿®æ”¹ã€‚å°±åƒåœ¨ JSX ä¸­çš„ props ä¸€æ ·ï¼Œå½“å€¼è¢«ä¼ é€’ç»™ Hook æ—¶ï¼Œå®ƒä»¬å°±åº”è¯¥æ˜¯ä¸å¯å˜çš„äº†ã€‚

```js {4}
function useIconStyle(icon) {
  const theme = useContext(ThemeContext);
  if (icon.enabled) {
    icon.className = computeStyle(icon, theme); // ğŸ”´ é”™è¯¯çš„ï¼šæ°¸è¿œä¸è¦ç›´æ¥ä¿®æ”¹ Hook çš„å‚æ•°ã€‚
  }
  return icon;
}
```

```js {3}
function useIconStyle(icon) {
  const theme = useContext(ThemeContext);
  const newIcon = { ...icon }; // âœ… æ­£ç¡®çš„ï¼šåˆ›å»ºä¸€ä¸ªæ–°çš„å‰¯æœ¬æ›¿ä»£
  if (icon.enabled) {
    newIcon.className = computeStyle(icon, theme);
  }
  return newIcon;
}
```

åœ¨ React ä¸­æœ‰ä¸€ä¸ªé‡è¦çš„åŸåˆ™å«åšå±€éƒ¨æ¨ç†ï¼Œå³é€šè¿‡å•ç‹¬æŸ¥çœ‹ç»„ä»¶æˆ– Hook çš„ä»£ç ï¼Œå°±èƒ½ç†è§£å®ƒçš„ä½œç”¨ã€‚å½“è°ƒç”¨ Hook æ—¶ï¼Œåº”è¯¥æŠŠå®ƒä»¬å½“ä½œâ€œé»‘ç›’å­â€ã€‚ä¾‹å¦‚ï¼Œè‡ªå®šä¹‰ Hook å¯èƒ½ä½¿ç”¨å…¶å‚æ•°ä½œä¸ºä¾èµ–é¡¹ï¼Œåœ¨å†…éƒ¨ç¼“å­˜å€¼ï¼š

```js {4}
function useIconStyle(icon) {
  const theme = useContext(ThemeContext);

  return useMemo(() => {
    const newIcon = { ...icon };
    if (icon.enabled) {
      newIcon.className = computeStyle(icon, theme);
    }
    return newIcon;
  }, [icon, theme]);
}
```

If you were to mutate the Hooks arguments, the custom hook's memoization will become incorrect,  so it's important to avoid doing that.

```js {4}
style = useIconStyle(icon);         // `style` æ˜¯åŸºäº `icon` è¿›è¡Œè®°å¿†åŒ–çš„
icon.enabled = false;               // é”™è¯¯çš„ï¼š ğŸ”´ æ°¸è¿œä¸è¦ç›´æ¥ä¿®æ”¹ Hook çš„å‚æ•°
style = useIconStyle(icon);         // è¿”å›ä¹‹å‰è®°å¿†åŒ–çš„ç»“æœ
```

```js {4}
style = useIconStyle(icon);         // `style` æ˜¯åŸºäº `icon` è¿›è¡Œè®°å¿†åŒ–çš„
icon = { ...icon, enabled: false }; // æ­£ç¡®çš„: âœ… åˆ›å»ºä¸€ä¸ªæ–°çš„å‰¯æœ¬æ›¿ä»£
style = useIconStyle(icon);         // è®¡ç®— `style` çš„æ–°å€¼
```

Similarly, it's important to not modify the return values of Hooks, as they may have been memoized.

---

## ä¸è¦æ”¹å˜ä¼ é€’ç»™ JSX åçš„å€¼ {/*values-are-immutable-after-being-passed-to-jsx*/}

ä¸è¦åœ¨ JSX ä½¿ç”¨è¿‡å€¼ä¹‹åæ”¹å˜å®ƒä»¬ã€‚åº”è¯¥åœ¨åˆ›å»º JSX ä¹‹å‰å®Œæˆå€¼çš„æ›´æ”¹ã€‚

å½“ä½ åœ¨è¡¨è¾¾å¼ä¸­ä½¿ç”¨ JSX æ—¶ï¼ŒReact å¯èƒ½ä¼šåœ¨ç»„ä»¶å®Œæˆæ¸²æŸ“ä¹‹å‰å°±æ€¥äºè®¡ç®— JSXã€‚è¿™æ„å‘³ç€ï¼Œå¦‚æœåœ¨å°†å€¼ä¼ é€’ç»™ JSX ä¹‹åå¯¹å®ƒä»¬è¿›è¡Œæ›´æ”¹ï¼Œå¯èƒ½ä¼šå¯¼è‡´ UI è¿‡æ—¶ï¼Œå› ä¸º React ä¸ä¼šçŸ¥é“éœ€è¦æ›´æ–°ç»„ä»¶çš„è¾“å‡ºã€‚

```js {4}
function Page({ colour }) {
  const styles = { colour, size: "large" };
  const header = <Header styles={styles} />;
  styles.size = "small"; // ğŸ”´ é”™è¯¯çš„ï¼šstyles å·²ç»åœ¨ä¸Šé¢çš„ JSX ä¸­ä½¿ç”¨äº†ã€‚
  const footer = <Footer styles={styles} />;
  return (
    <>
      {header}
      <Content />
      {footer}
    </>
  );
}
```

```js {4}
function Page({ colour }) {
  const headerStyles = { colour, size: "large" };
  const header = <Header styles={headerStyles} />;
  const footerStyles = { colour, size: "small" }; // âœ… æ­£ç¡®çš„ï¼šæˆ‘ä»¬åˆ›å»ºäº†ä¸€ä¸ªæ–°çš„å€¼ã€‚
  const footer = <Footer styles={footerStyles} />;
  return (
    <>
      {header}
      <Content />
      {footer}
    </>
  );
}
```
