---
title: Hook è§„åˆ™
---

<Intro>
Hook æ˜¯é€šè¿‡ JavaScript å‡½æ•°å®šä¹‰çš„ï¼Œä½†å®ƒä»¬ä»£è¡¨äº†ä¸€ç§ç‰¹æ®Šçš„å¯é‡ç”¨çš„ UI é€»è¾‘ï¼Œå¹¶ä¸”å¯¹å®ƒä»¬çš„è°ƒç”¨ä½ç½®æœ‰é™åˆ¶ã€‚
</Intro>

<InlineToc />

---

## åªåœ¨é¡¶å±‚è°ƒç”¨ Hook {/*only-call-hooks-at-the-top-level*/}

åç§°ä»¥ `use` å¼€å¤´çš„å‡½æ•°åœ¨ React ä¸­è¢«ç§°ä¸º **[Hook](/reference/react)**ã€‚

**ä¸è¦åœ¨å¾ªç¯ã€æ¡ä»¶ã€åµŒå¥—å‡½æ•°æˆ– `try`/`catch`/`finally` å—ä¸­è°ƒç”¨ Hook** ç›¸åï¼Œä½ åº”è¯¥åœ¨ React å‡½æ•°çš„é¡¶å±‚è°ƒç”¨ Hookï¼Œå¹¶ä¸”åœ¨ä»»ä½•æå‰è¿”å›ä¹‹å‰ã€‚ä½ åªèƒ½åœ¨ React æ¸²æŸ“å‡½æ•°ç»„ä»¶æ—¶è°ƒç”¨ Hookï¼š

* âœ… åœ¨[å‡½æ•°ç»„ä»¶](/learn/your-first-component)çš„é¡¶å±‚è°ƒç”¨ Hookã€‚
* âœ… åœ¨[è‡ªå®šä¹‰ Hook](/learn/reusing-logic-with-custom-hooks)çš„é¡¶å±‚è°ƒç”¨ Hookã€‚

```js{2-3,8-9}
function Counter() {
  // âœ… Good: åœ¨å‡½æ•°ç»„ä»¶é¡¶å±‚
  const [count, setCount] = useState(0);
  // ...
}

function useWindowWidth() {
  // âœ… Good: åœ¨è‡ªå®šä¹‰ Hooks é¡¶å±‚
  const [width, setWidth] = useState(window.innerWidth);
  // ...
}
```

åœ¨å…¶ä»–ä»»ä½•æƒ…å†µä¸‹è°ƒç”¨ä»¥ `use` å¼€å¤´çš„ Hook æ˜¯ä¸æ”¯æŒçš„ï¼Œä¾‹å¦‚ï¼š

* ğŸ”´ ä¸è¦åœ¨æ¡ä»¶æˆ–å¾ªç¯ä¸­è°ƒç”¨ Hookã€‚
* ğŸ”´ ä¸è¦åœ¨æ¡ä»¶ `return` è¯­å¥ä¹‹åè°ƒç”¨ Hookã€‚
* ğŸ”´ ä¸è¦åœ¨äº‹ä»¶å¤„ç†ç¨‹åºä¸­è°ƒç”¨ Hookã€‚
* ğŸ”´ ä¸è¦åœ¨ç±»ç»„ä»¶ä¸­è°ƒç”¨ Hookã€‚
* ğŸ”´ ä¸è¦åœ¨ä¼ é€’ç»™ `useMemo`ã€`useReducer` æˆ– `useEffect` çš„å‡½æ•°å†…éƒ¨è°ƒç”¨ Hookã€‚
* ğŸ”´ ä¸è¦åœ¨ `try`/`catch`/`finally` å—ä¸­è°ƒç”¨ Hookã€‚

å¦‚æœä½ è¿åäº†è¿™äº›è§„åˆ™ï¼Œä½ å¯èƒ½ä¼šçœ‹åˆ°ä»¥ä¸‹é”™è¯¯ï¼š

```js{3-4,11-12,20-21}
function Bad({ cond }) {
  if (cond) {
    // ğŸ”´ Bad: inside a condition (to fix, move it outside!)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad() {
  for (let i = 0; i < 10; i++) {
    // ğŸ”´ Bad: inside a loop (to fix, move it outside!)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad({ cond }) {
  if (cond) {
    return;
  }
  // ğŸ”´ Bad: after a conditional return (to fix, move it before the return!)
  const theme = useContext(ThemeContext);
  // ...
}

function Bad() {
  function handleClick() {
    // ğŸ”´ Bad: inside an event handler (to fix, move it outside!)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad() {
  const style = useMemo(() => {
    // ğŸ”´ Bad: inside useMemo (to fix, move it outside!)
    const theme = useContext(ThemeContext);
    return createStyle(theme);
  });
  // ...
}

class Bad extends React.Component {
  render() {
    // ğŸ”´ Bad: inside a class component (to fix, write a function component instead of a class!)
    useEffect(() => {})
    // ...
  }
}

function Bad() {
  try {
    // ğŸ”´ Bad: inside try/catch/finally block (to fix, move it outside!)
    const [x, setX] = useState(0);
  } catch {
    const [x, setX] = useState(1);
  }
}
```

ä½ å¯ä»¥ä½¿ç”¨ [`eslint-plugin-react-hooks` æ’ä»¶](https://www.npmjs.com/package/eslint-plugin-react-hooks) æ¥æ•æ‰è¿™äº›é”™è¯¯ã€‚

<Note>

[è‡ªå®šä¹‰ Hook](/learn/reusing-logic-with-custom-hooks) **å¯ä»¥** è°ƒç”¨å…¶ä»– Hookï¼ˆè¿™å°±æ˜¯å®ƒä»¬çš„ç›®çš„ï¼‰ã€‚å› ä¸ºè‡ªå®šä¹‰ Hook ä¹Ÿåªèƒ½åœ¨å‡½æ•°ç»„ä»¶æ¸²æŸ“æ—¶è¢«è°ƒç”¨ã€‚

</Note>

---

## åªåœ¨ React å‡½æ•°ä¸­è°ƒç”¨ Hook {/*only-call-hooks-from-react-functions*/}

ä¸è¦åœ¨å¸¸è§„çš„ JavaScript å‡½æ•°ä¸­è°ƒç”¨ Hookã€‚ç›¸åï¼Œä½ å¯ä»¥ï¼š

âœ… åœ¨ React å‡½æ•°ç»„ä»¶ä¸­è°ƒç”¨ Hookã€‚
âœ… åœ¨ [è‡ªå®šä¹‰ Hook](/learn/reusing-logic-with-custom-hooks#extracting-your-own-custom-hook-from-a-component) ä¸­è°ƒç”¨ Hookã€‚

éµå¾ªè¿™ä¸ªè§„åˆ™ï¼Œä½ å¯ä»¥ç¡®ä¿ç»„ä»¶ä¸­çš„æ‰€æœ‰æœ‰çŠ¶æ€é€»è¾‘åœ¨å…¶æºä»£ç ä¸­æ¸…æ™°å¯è§ã€‚

```js {2,5}
function FriendList() {
  const [onlineStatus, setOnlineStatus] = useOnlineStatus(); // âœ…
}

function setOnlineStatus() { // âŒ Not a component or custom Hook!
  const [onlineStatus, setOnlineStatus] = useOnlineStatus();
}
```
